vyper/venom/passes/loop_invariant_hosting.py

```diff
@@ -0,0 +1,127 @@
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.analysis.liveness import LivenessAnalysis
+from vyper.venom.analysis.loop_detection import LoopDetectionAnalysis
+from vyper.venom.basicblock import IRBasicBlock, IRInstruction, IRLiteral, IRVariable
+from vyper.venom.function import IRFunction
+from vyper.venom.passes.base_pass import IRPass
+
+def _ignore_instruction(instruction: IRInstruction) -> bool:
+    return (
+        instruction.is_volatile
+        or instruction.is_bb_terminator
+        or instruction.opcode == "returndatasize"
+        or instruction.opcode == "phi"
+    )
+
+
+def _is_correct_store(instruction: IRInstruction) -> bool:
+    return (
+        instruction.opcode == "store"
+        and len(instruction.operands) == 1
+        and isinstance(instruction.operands[0], IRLiteral)
+    )
+
+
+class LoopInvariantHoisting(IRPass):
+    """
+    This pass detects invariants in loops and hoists them above the loop body.
+    Any VOLATILE_INSTRUCTIONS, BB_TERMINATORS CFG_ALTERING_INSTRUCTIONS are ignored
+    """
+
+    from typing import Iterator
+
+    function: IRFunction
+    loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+    dfg: DFGAnalysis
+
+    def run_pass(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+        loops = self.analyses_cache.request_analysis(LoopDetectionAnalysis)
+        self.loops = loops.loops
+        while True:
+            change = False
+            for from_bb, loop in self.loops.items():
+                hoistable: list[
+                    tuple[IRBasicBlock, IRBasicBlock, IRInstruction]
+                ] = self._get_hoistable_loop(from_bb, loop)
+                if len(hoistable) == 0:
+                    continue
+                change |= True
+                self._hoist(hoistable)
+            if not change:
+                break
+            # I have this inside the loop because I dont need to
+            # invalidate if you dont hoist anything
+            self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+
+    def _hoist(self, hoistable: list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]]):
+        for loop_idx, bb, inst in hoistable:
+            bb.remove_instruction(inst)
+            bb_before: IRBasicBlock = loop_idx
+            bb_before.insert_instruction(inst, index=len(bb_before.instructions) - 1)
+
+    def _get_hoistable_loop(
+        self, from_bb: IRBasicBlock, loop: OrderedSet[IRBasicBlock]
+    ) -> list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]]:
+        result: list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]] = []
```

Comment:
these type signatures are hard to read -- either make them less specific (e.g. ` -> list` or `-> list[tuple]`) or make a small dataclass local to this file to carry around the data structure.